File: src/files/matting/background_negation_matter.cpp
#include "background_negation_matter.hpp"

#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>

cv::Mat BackgroundNegationMatter::background_mask(const cv::Mat &video_frame) {
  cv::Mat mask(video_frame.size(), video_frame.type());

  cv::absdiff(video_frame, background, mask);
  cv::cvtColor(mask, mask, cv::COLOR_BGR2GRAY);

  double threshold{stod(config.get(threshold_field))};
  double max_val{255.0};

  cv::threshold(mask, mask, threshold, max_val, cv::THRESH_BINARY);

  return mask;
}

File: kotlin_old/src/main/kotlin/uk/ac/ic/matterialize/matting/BackgroundNegationMatter.kt
package uk.ac.ic.matterialize.matting

import org.opencv.core.Core
import org.opencv.core.Mat
import org.opencv.imgproc.Imgproc
import org.opencv.imgproc.Imgproc.THRESH_BINARY

class BackgroundNegationMatter(private val background: Mat) : Matter {
    override fun backgroundMask(videoFrame: Mat): Mat {
        val mask = Mat()

        Core.absdiff(videoFrame, background, mask)
        Imgproc.cvtColor(mask, mask, Imgproc.COLOR_BGR2GRAY)

        val threshold = 25.0
        val maxval = 255.0
        Imgproc.threshold(mask, mask, threshold, maxval, THRESH_BINARY)

        return mask
    }
}
